### PR TITLE
clear all loader caches with new dir for plugin

### LINKED ISSUE #17078: customised module in role/library not being picked up in specific example
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

Use of custom modules in $role/library/ 
##### ANSIBLE VERSION

All ansible 2 versions including latest devel

```
ansible 2.2.0 (devel 375f7c515a) last updated 2016/08/09 14:32:21 (GMT +200)
  lib/ansible/modules/core: (detached HEAD 5d7b46e0dd) last updated 2016/08/13 17:11:47 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD eaa71f51d6) last updated 2016/08/13 17:11:47 (GMT +200)
  config file = 
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

None
##### OS / ENVIRONMENT

Runing Ansible on Ubuntu 16.04 controller
##### SUMMARY

Ansible will not pick up a customised module in role/library, given a specific set playbook setup. It will run the default module packaged with ansible instead.
##### STEPS TO REPRODUCE

See following links describing a minimal case that shows the problem:
- http://sprunge.us/dWXQ
- https://gist.github.com/srgvg/3a61651a98169e5bbb06a2dc726cbda5

Basically a small role that has a custom uri.py module and calls it.

I tracked the issue down, from a complex customer setup down to this minimal example.
When the play executing this role is prepended by a play, with some specific syntax, the error shows.

Issue still shows if the first play also targets `hosts: localhost` instead of being skipped.
Issue strangely disappears when the `name:` tag is deleted, leaving onky the plan `debug:` task
##### EXPECTED RESULTS

```
TASK [joeri : uri] *************************************************************
ok: [localhost] => {"changed": false, "failed": false, "msg": "custom module run"}
```
##### ACTUAL RESULTS

```
TASK [joeri : uri] *************************************************************
ok: [localhost] => {"accept_ranges": "none", "cache_control": "private, max-age=0", "changed": false, "connection": "close", "content_type": "text/html; charset=ISO-8859-1", "date": "Sat, 13 Aug 2016 15:12:05 GMT", "expires": "-1", "msg": "OK (unknown bytes)", "p3p": "CP=\"This is not a P3P policy! See https://www.google.com/support/accounts/answer/151657?hl=en for more info.\"", "redirected": true, "server": "gws", "set_cookie": "NID=84=eC_MViZ-Nx8Nf8XoDnAQAlDVW8uYu0cedyJwR4yCucx2-MF8f6iw_H2L5ZU7rGBUa_VFKPMT7WjX2yKWjKmbQBa9rf6SeBJgBV8GIF--Y3-q_jfoeUGJb3djGlNhELuS; expires=Sun, 12-Feb-2017 15:12:05 GMT; path=/; domain=.google.be; HttpOnly", "status": 200, "url": "http://www.google.be/?gfe_rd=cr&ei=xTivV8-wBZPDaL3kkegO", "vary": "Accept-Encoding", "x_frame_options": "SAMEORIGIN", "x_xss_protection": "1; mode=block"}
```

### PR DESCRIPTION
**Pull request by bcoca**
fixes #17078

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

plugin loader

### CODE REVIEW

File: changelogs/fragments/fix_plugin_loader_cache.yml [hunk 1]
```diff
@@ -0,0 +1,3 @@
+bugfixes:
+    - > clear all caches in plugin loader for a plugin type when adding new paths,
```
Comment (agaffney) [inline changelogs/fragments/fix_plugin_loader_cache.yml]: This doesn't seem to be valid YAML syntax

File: lib/ansible/plugins/loader.py [hunk 2]
```diff
@@ -269,16 +285,19 @@ def add_directory(self, directory, with_subdir=False):
...
             if with_subdir:
                 directory = os.path.join(directory, self.subdir)
             if directory not in self._extra_dirs:
                 # append the directory and invalidate the path cache
                 self._extra_dirs.append(directory)
-                self._paths = None
                 display.debug('Added %s to loader search path' % (directory))
 
+        if orig_size != len(self._extra_dirs):
+            self._clear_caches()
```
Comment (agaffney) [inline lib/ansible/plugins/loader.py]: Why not just do this unconditionally after the `append()` call above?

File: lib/ansible/plugins/loader.py [hunk 3]
```diff
@@ -113,11 +113,27 @@ def __init__(self, class_name, package, config, subdir, aliases=None, required_b
...
+        # hold dirs added at runtime outside of config
+        self._extra_dirs = []
+
+        # caches
         self._module_cache = MODULE_CACHE[class_name]
         self._paths = PATH_CACHE[class_name]
         self._plugin_path_cache = PLUGIN_PATH_CACHE[class_name]
 
-        self._extra_dirs = []
+        self._searched_paths = set()
```
Comment (agaffney) [inline lib/ansible/plugins/loader.py:124]: It seems like the above 5 lines or so could just be replaced with a call to `self._clear_caches()`

File: lib/ansible/plugins/loader.py (hunk 2 above)
Comment (bcoca) [reply lib/ansible/plugins/loader.py]: was thinking of accepting multiple dirs .. but that is probably premature

File: lib/ansible/plugins/loader.py (hunk 3 above)
Comment (bcoca) [reply lib/ansible/plugins/loader.py:124]: i dont want to clear the global caches in init

**Review by nitzmahone — COMMENTED** [review]
Given the potential for this to subtly change existing (and long-standing) behavior, and that this style of plugin loading may be marked "legacy" soon; I'm thinking we might want to just let sleeping dogs lie and skip this change altogether...

**Pull request merged**